File: week1_basic_data_structures/1_brackets_in_code/check_brackets.cpp
#include <iostream>
#include <stack>
#include <string>

struct Bracket {
    Bracket(char type, int position):
        type(type),
        position(position)
    {}

    bool Matchc(char c) {
        if (type == '[' && c == ']')
            return true;
        if (type == '{' && c == '}')
            return true;
        if (type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
};

int main() {
    std::string text;
    getline(std::cin, text);

    std::stack <Bracket> opening_brackets_stack;
    for (int position = 0; position < text.length(); ++position) {
        char next = text[position];

        if (next == '(' || next == '[' || next == '{') {
	  opening_brackets_stack.push(Bracket(next,position));
        }

        if (next == ')' || next == ']' || next == '}') {
	  if(opening_brackets_stack.size() == 0 || !(opening_brackets_stack.top().Matchc(next)) ){
	    std::cout << position+1 << "\n";
	    return 0;
	  }
	  else{
	    opening_brackets_stack.pop();
	  }
	}
    }
    if(opening_brackets_stack.size() == 0){
      	    std::cout <<  "Success\n";
    }
    else{
      std::cout << opening_brackets_stack.top().type <<"\n";
    }


    // Printing answer, write your code here

    return 0;
}

File: week1_basic_data_structures/1_brackets_in_code/check_brackets.py
# python3

from collections import namedtuple

Bracket = namedtuple("Bracket", ["char", "position"])


def are_matching(left, right):
    return (left + right) in ["()", "[]", "{}"]


def find_mismatch(text):
    opening_brackets_stack = []
    for i, next in enumerate(text):
        if next in "([{":
            opening_brackets_stack.append(Bracket(next,i))
            # Process opening bracket, write your code here

        if next in ")]}":
            if(len(opening_brackets_stack) == 0 or not are_matching(opening_brackets_stack[-1][0],next)):
                return i+1
            else:
                opening_brackets_stack.pop()
            # Process closing bracket, write your code here
    if(len(opening_brackets_stack) != 0):
        return opening_brackets_stack[0][1]+1
    else:
        return "Success"

def main():
    text = input()
    mismatch = find_mismatch(text)
    # Printing answer, write your code here
    print(mismatch)

if __name__ == "__main__":
    main()
